### PR TITLE
Moving Average Tracking

### PR DESCRIPTION
**Pull request by npritch928**
This commit was meant to incorporate the tracking methodology from "Towards Practical Large-Scale Randomized Iterative Least Squares Solvers through Uncertainty Quantification" and "Solving, Tracking and Stopping Streaming Linear Inverse Problems" by Pritchard and Patel into the library. This tracking method allows for the random blocks to be used for tracking rather than needing to access the entire matrix. The changes that were made to accomplish this were:

1. `LSLogMA` is a log to store the tracking information.
2. `LSStopMA` a stopping criterion for the moving average.
3. A function `get_uncertainty` which computes the uncertainty sets.
4. A function `get_SE_constants!` computes sub-Exponential  and  parameters for the defined sampling methods. I have not yet done this for the uniform method because I need to still run experiments to determine what the values should be.
5. `LSStopThreshold` which stops the solver when a specific threshold is achieved.
6. Tests for all of these types.
7. An example for the moving average tracking.

### CODE REVIEW

**Comment by dmaldona** [comment]
@npritch928 @vp314 I think this would benefit of a new entry in the Manual (Documenter.jl).

File: docs/src/api/linear_solver_logs.md [hunk 1]
```diff
@@ -17,6 +17,8 @@ LinSysSolverLog
 LSLogOracle
 
 LSLogFull
+
+LSLogFullMA
```
Comment (vp314) [inline docs/src/api/linear_solver_logs.md:21]: Does this need to be called "Full"? Maybe change it in the future

File: src/linear_solver_logs/solve_log_full_ma.jl [hunk 2]
```diff
@@ -0,0 +1,248 @@
+# This file is part of RLinearAlgebra.jl
+
+# using LinearAlgebra
+
+mutable struct MAInfo
```
Comment (vp314) [inline src/linear_solver_logs/solve_log_full_ma.jl:5]: This needs a doc string

File: src/linear_solver_logs/solve_log_full_ma.jl [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+- `true_res`, a boolean indicating if we want the true residual computed instead of approximate.
+# Constructors
+- Calling `LSLogFullMA()` sets `collection_rate = 1`,  `lambda1 = 1`,
+    `lambda2 = 30`, `resid_hist = Float64[]`, `iota_hist = Float64[]`, `width_hist = Int64[]`, 
+    `resid_norm = norm` (Euclidean norm), `iterations = -1`, `converged = false`, 
+    `sampler = LinSysVecRowDetermCyclic`, `max_dimension = 0`, `sigma2 = nothing`, `omega = nothing`,
+    `eta = 1`, and `true_res = false`. The user can specify their own values of lambda1, lambda2, sigma2, omega, and eta using 
+    key word arguments as inputs into the constructor.
+"""
+mutable struct LSLogFullMA <: LinSysSolverLog
```
Comment (vp314) [inline src/linear_solver_logs/solve_log_full_ma.jl:59]: It is not clear to me that all of these fields are needed. Is some of this information not stored in MAInfo or can be determined from MAInfo?

File: src/linear_solver_logs/solve_log_full_ma.jl [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+
+        end
+
+    end
+
+end
+
+# Update the moving average estimator requires the log variable, ma_info,
+# observed residual, and a lambda_base which corresponds to which lambda regime we are in. 
+function Update_MAEstimators!(log, ma_info, res, lambda_base, iter)
```
Comment (vp314) [inline src/linear_solver_logs/solve_log_full_ma.jl:154]: Does this follow the BLUE style convention?

Comment (vp314) [reply src/linear_solver_logs/solve_log_full_ma.jl:154]: Also, why are we not passing types here?

File: src/linear_solver_logs/solve_log_full_ma.jl [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+        
+        # Determine start point for first loop
+        startp1 = diff < 0 ? 1 : (diff + 1)
+        
+        # Determine start and endpoints for second loop
+        startp2 = diff > 0 ? 2 : lambda_base + diff + 1  
+        endp2 = diff > 0 ? 1 : lambda_base
+
+        # Compute the moving average two loop setup required when lambda < lambda2
+        @simd for i in startp1:ma_info.idx
```
Comment (vp314) [inline src/linear_solver_logs/solve_log_full_ma.jl:185]: Do we want to use these simd commands right now?

File: src/linear_solver_logs/solve_log_full_ma.jl (hunk 4 above)
Comment (vp314) [reply src/linear_solver_logs/solve_log_full_ma.jl:154]: Docstring?

File: src/linear_solver_stops/stop_ma_thres.jl [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+mutable struct LSStopMA <: LinSysStopCriterion
+    max_iter::Int64
+    threshold::Float64
+    delta1::Float64
+    delta2::Float64
+    chi1::Float64
+    chi2::Float64
+end
+
+LSStopMA(iter;
```
Comment (vp314) [inline src/linear_solver_stops/stop_ma_thres.jl:27]: doctstring does not include constructor?

File: src/linear_solver_logs/solve_log_full_ma.jl [hunk 7]
```diff
@@ -0,0 +1,248 @@
...
+        
+        ma_info.lambda += 1
+    end
+
+end
+
+#Function that will return rho and its uncertainty from a LSLogFullMA type 
+"""
+    get_uncertainty(log::LSLogFullMA; alpha = .95)
+    A function that takes a LSLogFullMA type and a confidence level, alpha, and returns credible intervals for for every rho in the log, specifically it returns a tuple with (rho, Upper bound, Lower bound).
```
Comment (vp314) [inline src/linear_solver_logs/solve_log_full_ma.jl:210]: Should there be a blank line above this and this perhaps should not be indented?

**Review by vp314 — CHANGES_REQUESTED** [review]
See individual comments in the changed files

File: docs/src/api/linear_solver_logs.md (hunk 1 above)
Comment (npritch928) [reply docs/src/api/linear_solver_logs.md:21]: I removed the 'full' name and updated all relevant files.

File: src/linear_solver_logs/solve_log_full_ma.jl (hunk 4 above)
Comment (npritch928) [reply src/linear_solver_logs/solve_log_full_ma.jl:154]: I believe we have been following a convention that if a function is not exported no docstrings are added because the user should not be using non-exported functions. I added the types and updated the name to align with BLUE

File: src/linear_solver_logs/solve_log_full_ma.jl (hunk 5 above)
Comment (npritch928) [reply src/linear_solver_logs/solve_log_full_ma.jl:185]: I removed the simd

File: src/linear_solver_stops/stop_ma_thres.jl [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+        thresholdChecks = sqrt(log.iota_hist[its]) <= I_threshold && 
+        log.resid_hist[its] <= stop.threshold
+    else
+        thresholdChecks = false
+    end
+    return (thresholdChecks || its == stop.max_iter ? true : false)
+end
+
+# Once the sigma2 is known function computest the threshold
+function iota_threshold(hist::LSLogFullMA, stop::LSStopMA)
```
[the remark this answers is not in the document]
Comment (npritch928) [reply src/linear_solver_stops/stop_ma_thres.jl:52]: As I noted earlier this function is for internal use only, so I did not include a doctstring because the users should not be utilizing this function.

File: src/linear_solver_stops/stop_ma_thres.jl (hunk 6 above)
Comment (npritch928) [reply src/linear_solver_stops/stop_ma_thres.jl:27]: I added the constructor.

File: src/linear_solver_logs/solve_log_full_ma.jl (hunk 7 above)
Comment (npritch928) [reply src/linear_solver_logs/solve_log_full_ma.jl:210]: I made these changes

File: src/linear_rsolve.jl [hunk 9]
```diff
@@ -108,8 +108,8 @@ function rsolve!(solver::RLSSolver, A, b, x::AbstractVector)
     while check_stop_criterion(solver.log, solver.stop) == false
         iter += 1
         samp = sample(solver.sampler, A, b, x, iter)
-        rsubsolve!(solver.routine, x, samp, iter)
         log_update!(solver.log, solver.sampler, x, samp, iter, A, b)
+        rsubsolve!(solver.routine, x, samp, iter)
```
Comment (dmaldona) [inline src/linear_rsolve.jl:112]: @npritch928 why was this needed?

File: src/linear_solver_logs/solve_log_full_ma.jl [hunk 10]
```diff
@@ -0,0 +1,329 @@
...
+                    Float64[], 
+                    Float64[], 
+                    Int64[],
+                    norm, 
+                    -1, 
+                    false,
+                    false,
+                    DistInfo(nothing, 0, 0, nothing, nothing, 1)
+                    )
+
```
Comment (dmaldona) [inline src/linear_solver_logs/solve_log_full_ma.jl:112]: @npritch928 can you check this is the right way to define the default constructor? Look at the compilation warning

`  1 dependency had warnings during precompilation:
┌ RLinearAlgebra [da1d211a-c3bc-43b2-b5d0-ce83f16a0f3a]
│  WARNING: Method definition (::Type{RLinearAlgebra.LSLogMA})() in module RLinearAlgebra at /home/adrian/Project/RLinearAlgebra.jl/src/linear_solver_logs/solve_log_full_ma.jl:100 overwritten at /home/adrian/Project/RLinearAlgebra.jl/src/linear_solver_logs/solve_log_full_ma.jl:113.
│    ** incremental compilation may be fatally broken for this module **`

File: src/linear_rsolve.jl (hunk 9 above)
Comment (npritch928) [reply src/linear_rsolve.jl:112]: If we use sketch $(\tilde A_k,$\tilde b_k)$ the update would be $x_k = x_{k-1} - \tilde A_k^\top (\tilde A_k \tilde A_k)^\dagger (\tilde A_k x_{k-1} - \tilde b_k)$, if we compute the sketched residual of that value using the same sketch, which would have happened when rsubsolve is before log_update, this residual will always be zero. $\tilde A_k x_{k-1} - \tilde A_k \tilde A_k^\top (\tilde A_k \tilde A_k)^\dagger (\tilde A_k x_{k-1} - \tilde b_k) - \tilde b_k = 0$. So the order needed to be changed

File: src/linear_solver_logs/solve_log_full_ma.jl (hunk 10 above)
Comment (npritch928) [reply src/linear_solver_logs/solve_log_full_ma.jl:112]: I checked it, it was an issue of me thinking the two constructors were different when they were in fact the same constructor. I used the one with the optional arguments.

File: docs/src/man/tracking_overview.md [hunk 11]
```diff
@@ -0,0 +1,153 @@
...
+# Generate a system
+A = rand(20, 5);
+x = rand(5);
+b = A * x;
+
+# Specify solver
+solver = RLSSolver(
+    LinSysVecRowRandCyclic(),   # Random Cyclic Sampling
+    LinSysVecRowProjStd(),      # Hyperplane projection
+    LSLogMA(lambda_2 = 100),# Full Logger: maintains moving average residual history
```
[the remark this answers is not in the document]
Comment (npritch928) [reply docs/src/man/tracking_overview.md:88]: done

File: docs/src/man/tracking_overview.md [hunk 12]
```diff
@@ -0,0 +1,153 @@
...
+These cost could be avoided by computing the sketched
+residual, i.e. $\|S A x - Sb\|_2^2$, but at the cost exact knowledge of 
+progress. To reduce this randomness Pritchard and Patel propose using 
+a moving average of the sketched residuals in "Solving, Tracking 
+and Stopping Streaming Linear Inverse Problems." Specifically they define
+the progress estimator 
+$$
+    \hat \rho_k^\lambda = \sum_{i-\lambda +1}^k \frac{\|S_i A x - S_ib\|_2^2}{\lambda},
+$$  
+where $\lambda$ is the width of the moving average window.
```
[the remark this answers is not in the document]
Comment (npritch928) [reply docs/src/man/tracking_overview.md:23]: I added a url link to the arXiv article

File: docs/src/man/tracking_overview.md [hunk 13]
```diff
@@ -0,0 +1,153 @@
...
+# Generate a system
+A = rand(20, 5);
+x = rand(5);
+b = A * x;
+
+# Specify solver
+solver = RLSSolver(
+    LinSysVecRowRandCyclic(),   # Random Cyclic Sampling
+    LinSysVecRowProjStd(),      # Hyperplane projection
+    LSLogMA(),              # Full Logger: maintains moving average residual history
```
[the remark this answers is not in the document]
Comment (npritch928) [reply docs/src/man/tracking_overview.md:40]: done

File: src/linear_solver_logs/solve_log_full_ma.jl (hunk 3 above)
Comment (npritch928) [reply src/linear_solver_logs/solve_log_full_ma.jl]: I added an extra type called dist_info to reduce the number of fields.

**Comment by vp314** [comment]
@npritch928  I would recommend updating the PR to reflect the name changes that (I believe) were made in the codebase.

**Review by vp314 — APPROVED** [review]
Overall, I think the in-line comments for `LSLogMA` need to be modified to remove the "Full" part. I think we should reserve full when the whole residual is calculated. If you disagree, then that is fine. I approved the merge.

**Review by dmaldona — APPROVED** [review]
